File: python/tree/BinaryTreeZigzagLevelOrderTraversal.py
# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of lists of integers
    def zigzagLevelOrder(self, root):
        result = []
        if root is None:
            return result
        
        left_to_right = True
        curr_lv = [root]
        while curr_lv:
            curr_line = []
            next_lv = []
            if left_to_right is False:
                for i in xrange(len(curr_lv)-1, -1, -1):
                    curr_line.append(curr_lv[i].val)
                    if curr_lv[i].right:
                        next_lv.append(curr_lv[i].right)
                    if curr_lv[i].left:
                        next_lv.append(curr_lv[i].left)
            else:
                for i in xrange(len(curr_lv)-1, -1, -1):
                    curr_line.append(curr_lv[i].val)
                    if curr_lv[i].left:
                        next_lv.append(curr_lv[i].left)
                    if curr_lv[i].right:
                        next_lv.append(curr_lv[i].right)
            result.append(curr_line)
            curr_lv = next_lv
            left_to_right = not left_to_right
        return result
                    
if __name__ == "__main__":
    solution = Solution()
    root = TreeNode(1)
    node2 = TreeNode(2)
    node3 = TreeNode(3)
    node4 = TreeNode(4)
    node5 = TreeNode(5)
    root.left = node2
    root.right = node3
    node2.left = node4
    node2.right = node5

    print solution.zigzagLevelOrder(root)

File: java/BinaryTreeZigzagLevelOrderTraversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        Boolean leftToRight = true;
        List<TreeNode> curr = new ArrayList<TreeNode>();
        if (root == null) {
            return result;
        }
        curr.add(root);
        while (curr.size() != 0) {
            List<Integer> curr_line = new ArrayList<Integer>();
            List<TreeNode> next = new ArrayList<TreeNode>();
            if (leftToRight == true) {
                for (int i = curr.size()-1; i>=0; i--) {
                    if (curr.get(i).left != null) {
                        next.add(curr.get(i).left);
                    }
                    if (curr.get(i).right != null) {
                        next.add(curr.get(i).right);
                    }
                    curr_line.add(curr.get(i).val);
                }
            } else {
                for (int i = curr.size()-1; i >= 0; i--) {
                    if (curr.get(i).right != null) {
                        next.add(curr.get(i).right);
                    }
                    if (curr.get(i).left != null) {
                        next.add(curr.get(i).left);
                    }
                    curr_line.add(curr.get(i).val);
                }
            }
            result.add(curr_line);
            leftToRight = !leftToRight;
            curr = next;
        }
        return result;
    }
}
